### PR TITLE
feat: add flags to control console logs

### PR DESCRIPTION
**Pull request by Romsters**
# What :computer: 
## Anvil flags:
- added `--silent`/`--quite` flag by analogy with anvil;
- added `--disable-console-log`/`--no-console-log` flag by analogy with anvil;

## Custom flags:
- added `--show-node-config` flag to control printing of node config on startup.
- added `--show-calls-summary` flag to control printing of calls and txs summary.
- added `--show-event-logs` flag to control printing of events.

- added missing default values to make flags like  `--resolve-hashes` work without specifying a value;

# Why :hand:
To be able to use the node with hardhat tests/scripts the only thing that has to be printed is contracts `console.log`.
This PR adds flags to control each section of logs. To run with hardhat tests/scripts use `--show-node-config=false --show-calls-summary=false --show-event-logs=false`.

### CODE REVIEW

File: src/node/in_memory.rs [hunk 1]
```diff
@@ -1171,34 +1171,39 @@ impl<S: ForkSource + std::fmt::Debug + Clone> InMemoryNode<S> {
...
+                ExecutionResult::Revert { output } => {
+                    tracing::info!("Call: {}: {}", "FAILED".red(), output);
+                }
+                ExecutionResult::Halt { reason } => {
+                    tracing::info!("Call: {} {}", "HALTED".red(), reason)
+                }
+            };
+            // Don't print the header if only contract calls should be shown
+            tracing::info!("=== Console Logs: ");
+        }
```
Comment (dutterbutter) [inline src/node/in_memory.rs:1190]: I dont think we should disable informing on reverts, halts, successes via this `show_only_contract_logs`. This flag should only toggle console logs, no?

File: src/config/cli.rs [hunk 2]
```diff
@@ -87,11 +87,15 @@ pub struct Cli {
...
 
-    #[arg(long, help_heading = "Debugging Options")]
+    #[arg(long, default_missing_value = "true", num_args(0..=1), help_heading = "Debugging Options")]
     /// If true, the tool will try to resolve ABI and topic names for better readability.
     /// May decrease performance.
     pub resolve_hashes: Option<bool>,
 
+    #[arg(long, default_missing_value = "true", num_args(0..=1), help_heading = "Debugging Options")]
+    /// If true, the tool will only show contract logs.
+    pub show_only_contract_logs: Option<bool>,
```
Comment (dutterbutter) [inline src/config/cli.rs:97]: I don't particularly understand the purpose of `show_only_contract_logs`. Why does hardhat need all other logs to be silenced? 

Ideally we shouldn't use `show_only_contract_logs` to conditionally enable / disable all other logs. This is what `quiet` should do. If Hardhat absolutely needs ALL other logs to be silenced then it should just use `--quiet`. We can then by default have console logs ignore this flag, and be used alongside `--disable-console-log` if devs want complete silence. As they also have:

```
--disable-console-log
   Disable printing of `console.log` invocations to stdout
    
   [aliases: no-console-log]
```

File: src/config/cli.rs [hunk 3]
```diff
@@ -139,6 +143,10 @@ pub struct Cli {
     /// Log file path (default: era_test_node.log).
     pub log_file_path: Option<String>,
 
+    #[arg(long, default_missing_value = "true", num_args(0..=1), help_heading = "Logging Configuration")]
+    /// If true, the tool will not print anything on startup.
+    pub silent: Option<bool>,
```
Comment (dutterbutter) [inline src/config/cli.rs:148]: This is now `quiet` in anvil:

```
  -q, --quiet
          Do not print log messages
```

File: src/node/in_memory.rs (hunk 1 above)
Comment (dutterbutter) [reply src/node/in_memory.rs:1190]: See point made above

File: src/node/in_memory.rs [hunk 4]
```diff
@@ -1171,34 +1171,39 @@ impl<S: ForkSource + std::fmt::Debug + Clone> InMemoryNode<S> {
...
-        tracing::info!("=== Call traces:");
-        for call in &call_traces {
-            formatter::print_call(
-                call,
-                0,
-                &inner.config.show_calls,
-                inner.config.show_outputs,
-                inner.config.resolve_hashes,
-            );
+        if !inner.config.show_only_contract_logs {
```
Comment (dutterbutter) [inline src/node/in_memory.rs:1196]: See point made above.

File: src/node/in_memory.rs [hunk 5]
```diff
@@ -1463,85 +1468,94 @@ impl<S: ForkSource + std::fmt::Debug + Clone> InMemoryNode<S> {
         let spent_on_pubdata =
             tx_result.statistics.gas_used - tx_result.statistics.computational_gas_used as u64;
 
-        tracing::info!("┌─────────────────────────┐");
-        tracing::info!("│   TRANSACTION SUMMARY   │");
-        tracing::info!("└─────────────────────────┘");
+        if !inner.config.show_only_contract_logs {
```
Comment (dutterbutter) [inline src/node/in_memory.rs:1471]: See point made above.

File: src/node/in_memory.rs [hunk 6]
```diff
@@ -1463,85 +1468,94 @@ impl<S: ForkSource + std::fmt::Debug + Clone> InMemoryNode<S> {
...
-        if inner.config.show_calls != ShowCalls::None {
-            for call in call_traces {
-                formatter::print_call(
-                    call,
-                    0,
-                    &inner.config.show_calls,
-                    inner.config.show_outputs,
-                    inner.config.resolve_hashes,
-                );
+        if !inner.config.show_only_contract_logs {
```
Comment (dutterbutter) [inline src/node/in_memory.rs:1527]: See point made above.

**Review by dutterbutter — CHANGES_REQUESTED** [review]
I think this should be refactored in favour of the proposed approach of using `--quiet` and `--disable-console-logs`. 

On a side note, thinking about our convo earlier, we can remove the need for all `show-x` flags in favour of verbosity flag as they have in `anvil`. That is:

```
  -v, --verbosity...
          Verbosity level of the log messages.
          
          Pass multiple times to increase the verbosity (e.g. -v, -vv, -vvv).
          
          Depending on the context the verbosity levels have different meanings.
          
          For example, the verbosity levels of the EVM are:
          - 2 (-vv): Print logs for all tests.
          - 3 (-vvv): Print execution traces for failing tests.
          - 4 (-vvvv): Print execution traces for all tests, and setup traces for failing tests.
          - 5 (-vvvvv): Print execution and setup traces for all tests, including storage changes.
```

For us we could modify the levels to tailor our needs for system / user calls, gas details, vm results, storage writes etc. I can create an issue for this to be discussed further.

File: src/config/cli.rs (hunk 3 above)
Comment (Romsters) [reply src/config/cli.rs:148]: wdym? I don't see `--quiet` in anvil, it's `--silent`

**Comment by Romsters** [comment]
> I think this should be refactored in favour of the proposed approach of using `--quiet` and `--disable-console-logs`.
> 
> On a side note, thinking about our convo earlier, we can remove the need for all `show-x` flags in favour of verbosity flag as they have in `anvil`. That is:
> 
> ```
>   -v, --verbosity...
>           Verbosity level of the log messages.
>           
>           Pass multiple times to increase the verbosity (e.g. -v, -vv, -vvv).
>           
>           Depending on the context the verbosity levels have different meanings.
>           
>           For example, the verbosity levels of the EVM are:
>           - 2 (-vv): Print logs for all tests.
>           - 3 (-vvv): Print execution traces for failing tests.
>           - 4 (-vvvv): Print execution traces for all tests, and setup traces for failing tests.
>           - 5 (-vvvvv): Print execution and setup traces for all tests, including storage changes.
> ```
> 
> For us we could modify the levels to tailor our needs for system / user calls, gas details, vm results, storage writes etc. I can create an issue for this to be discussed further.

Was thinking about that a lot, what I definitely don't want to do is printing anything when `--silent` is used. That's how anvil works. We should rework all our flags 100%, just didn't want it to be included in the scope of this PR. I don't see anvil having verbosity flag though.

**Comment by dutterbutter** [comment]
@Romsters the documentation is outdated. You need to build from source to view most recent flags. I agree we need to rework our flags, but I still the mentioned approach is better.

**Comment by Romsters** [comment]
> @Romsters the documentation is outdated. You need to build from source to view most recent flags. I agree we need to rework our flags, but I still think the mentioned approach is better.

Indeed, there quite a few new flags not yet available in their latest version, will review them.

File: src/config/cli.rs [hunk 7]
```diff
@@ -66,12 +66,28 @@ pub struct Cli {
     /// Enable default settings for debugging contracts.
     pub debug_mode: bool,
 
+    #[arg(long, default_value = "true", default_missing_value = "true", num_args(0..=1), help_heading = "Debugging Options")]
+    /// If true, prints node config on startup.
+    pub show_node_config: Option<bool>,
```
Comment (dutterbutter) [inline src/config/cli.rs:71]: Why do we need this? This should always show unless silent is used?

File: src/config/cli.rs [hunk 8]
```diff
@@ -66,12 +66,28 @@ pub struct Cli {
...
+
+    #[arg(long, default_value = "true", default_missing_value = "true", num_args(0..=1), help_heading = "Debugging Options")]
+    /// If true, prints calls and transactions summary.
+    pub show_calls_summary: Option<bool>,
```
Comment (dutterbutter) [inline src/config/cli.rs:75]: Perhaps rename this as its too close to `show_calls` while this flag is for transaction summary / outcome

File: src/config/cli.rs [hunk 9]
```diff
@@ -66,12 +66,28 @@ pub struct Cli {
...
+
+    #[arg(long, alias = "no-console-log", default_missing_value = "true", num_args(0..=1), help_heading = "Debugging Options")]
+    /// Disables printing of `console.log` invocations to stdout if true.
+    pub disable_console_log: Option<bool>,
+
+    #[arg(long, default_value = "true", default_missing_value = "true", num_args(0..=1), help_heading = "Debugging Options")]
+    /// If true, logs events.
+    pub show_event_logs: Option<bool>,
```
Comment (dutterbutter) [inline src/config/cli.rs:83]: Lets make this default false.

**Review by dutterbutter — CHANGES_REQUESTED** [review]

File: src/config/cli.rs (hunk 7 above)
Comment (Romsters) [reply src/config/cli.rs:71]: No, we don’t need it when using with hardhat tests and scripts. We only need console.log in that case

File: src/config/cli.rs (hunk 8 above)
Comment (Romsters) [reply src/config/cli.rs:75]: Thought about that, but I would better rename show_calls as it is not clear. The flag I added is for both eth_call and transactions.

File: src/config/cli.rs (hunk 9 above)
Comment (Romsters) [reply src/config/cli.rs:83]: Don’t mind

File: src/config/cli.rs (hunk 8 above)
Comment (Romsters) [reply src/config/cli.rs]: Renamed to show_tx_summary

File: src/config/cli.rs (hunk 9 above)
Comment (Romsters) [reply src/config/cli.rs]: done

File: src/node/in_memory.rs (hunk 1 above)
Comment (Romsters) [reply src/node/in_memory.rs]: Removed show_only_contract_logs and added individual flags to control every section of logs

**Review by dutterbutter — APPROVED** [review]

**Pull request merged**